### PR TITLE
[AMORO-2] Flink support for Setting/Altering Amoro table properties

### LINKED ISSUE #2: [Improvement][Flink] Adding support for setting/altering Arctic table properties
### Search before asking

- [x] I have searched in the [[issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue)](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?
Enhancing the ability of Flink SQL to operate Arctic tables, allowing users to modify table properties through the `alter table` syntax.
Set or Alter Table Properties
```sql
ALTER TABLE [catalog_name.][db_name.]table_name SET (key1=val1, key2=val2, ...)
```
Set one or more properties in the specified table. If a particular property is already set in the table, override the old value with the new one.

### How should we improve?

-  Implement the `ArcticCatalog.alterTable` method.
-  Invoking the AMS API to alter the properties.
-  Add some UTs with the Flink Catalog API and Flink SQL.
-  Affected Flink versions: flink1.12/flink1.14/flink1.15.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [[Code of Conduct](https://www.apache.org/foundation/policies/conduct)](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by czy006**
## Why are the changes needed?

Close #2.

## Brief change log

- Flink support for Setting/Altering Amoro table properties
- Add Tests about change the key properties

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible
- [ ] Add screenshots for manual tests if appropriate
- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java [hunk 1]
```diff
@@ -572,4 +572,84 @@ private void insertValue() {
 
     sql("DROP TABLE default_catalog.default_database." + TABLE);
   }
+
+  @Test
+  public void testAlterUnKeyTable() throws Exception {
```
Comment (majin1102) [inline mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java:577]: Why this is a legacy test?

**Review by majin1102 — CHANGES_REQUESTED** [review]
Thanks for your contribution.

I wonder why use a legacy test to support this new feature, which doesn't quite make sense

File: mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/MixedCatalog.java [hunk 2]
```diff
@@ -623,4 +651,131 @@ private void validateColumnOrder(CatalogBaseTable table) {
...
+
+    for (Map.Entry<String, String> entry : newTable.getOptions().entrySet()) {
+      String key = entry.getKey();
+      String value = entry.getValue();
+
+      if (Objects.equal(value, oldProperties.get(key))) {
+        continue;
+      }
+
+      if ("location".equalsIgnoreCase(key)) {
```
Comment (YesOrNo828) [inline mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/MixedCatalog.java:703]: I would prefer to define the string constants: "location", "current-snapshot-id", and "cherry-pick-snapshot-id". WDYT?

File: mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/MixedCatalog.java [hunk 3]
```diff
@@ -623,4 +651,131 @@ private void validateColumnOrder(CatalogBaseTable table) {
...
+      if (!Objects.equal(value, oldProperties.get(key))) {
+        setProperties.put(key, value);
+      }
+    }
+    oldProperties
+        .keySet()
+        .forEach(
+            k -> {
+              if (!newTable.getOptions().containsKey(k)) {
+                setProperties.put(k, null);
```
Comment (YesOrNo828) [inline mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/MixedCatalog.java:757]: Why setting a null value for the keys which doesn't exists in the newTable.Options?
```
tbl properties (
    k1 = v1
);
alter table tbl set k2=v2 where ....;
```
I think the k1-v1 property would be overrided after executing the ALTER statement.

**Review by YesOrNo828 — COMMENTED** [review]
@czy006
Thanks for the contribution. In general speaking the code is good for me. And I left some comments. PTAL

File: mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java (hunk 1 above)
Comment (czy006) [reply mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java:577]: #2344 Rename By @xieyi888 Maybe merge to TestMixedCatalogTablePartitions or future test?

File: mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/MixedCatalog.java (hunk 3 above)
Comment (czy006) [reply mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/MixedCatalog.java:757]: [FlinkAlterTableUtil.commitChanges](https://github.com/apache/iceberg/blob/main/flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/util/FlinkAlterTableUtil.java#L56) before commit will be remove

File: mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java (hunk 1 above)
Comment (majin1102) [reply mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java:577]: ye, I see
No problem

**Review by majin1102 — APPROVED** [review]
LGTM

File: mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java (hunk 1 above)
Comment (xieyi888) [reply mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java:577]: It seems a little confusion
As I know, LEGACY Type is for this case
```
CREATE CATALOG <catalog_name> WITH (
  'type'='arctic',
); 
```

we will have four catalog type in the future
https://github.com/NetEase/amoro/blob/d0f613fc26aafded9f40ed489a8df3dce4dfb86a/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/factories/CatalogFactoryOptions.java#L34
image

So we have Three Test case:

- TestLegacyMixedCatalog (tset case for type='arctic')

- TestMixedCatalog containe (tset case for type='mixed_iceberg' , type='mixed_hive')

- FlinkUnifiedCatalogITCase (tset case for type='unified')

But now, some test cases for type='mixed_iceberg' type='mixed_hive' is below TestLegacyMixedCatalog. It make confusion.
https://github.com/NetEase/amoro/blob/d0f613fc26aafded9f40ed489a8df3dce4dfb86a/mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java#L111
image

Maybe we could do some adjust in the future and make TestCase for different Catalog type more clearly @YesOrNo828 HDYT

**Comment by xieyi888** [comment]
> Thanks for your contribution.
> 
> I wonder why use a legacy test to support this new feature, which doesn't quite make sense

It seems a little confusion
As I know, LEGACY Type is for this case
```
CREATE CATALOG <catalog_name> WITH (
  'type'='arctic',
); 
```

we will have four catalog type in the future
https://github.com/NetEase/amoro/blob/d0f613fc26aafded9f40ed489a8df3dce4dfb86a/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/factories/CatalogFactoryOptions.java#L34
image

So we have Three Test case:

- TestLegacyMixedCatalog (tset case for type='arctic')

- TestMixedCatalog containe (tset case for type='mixed_iceberg' , type='mixed_hive')

- FlinkUnifiedCatalogITCase (tset case for type='unified')

But now, some test cases for type='mixed_iceberg' type='mixed_hive' is below TestLegacyMixedCatalog. It make confusion.
https://github.com/NetEase/amoro/blob/d0f613fc26aafded9f40ed489a8df3dce4dfb86a/mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java#L111
image

Maybe we could do some adjust in the future and make TestCase for different Catalog type more clearly @YesOrNo828 HDYT

**Review by YesOrNo828 — APPROVED** [review]
LGTM

**Comment by czy006** [comment]
> > Thanks for your contribution.
> > I wonder why use a legacy test to support this new feature, which doesn't quite make sense
> 
> It seems a little confusion As I know, LEGACY Type is for this case
> 
> ```
> CREATE CATALOG <catalog_name> WITH (
>   'type'='arctic',
> ); 
> ```
> 
> we will have four catalog type in the future
> 
> https://github.com/NetEase/amoro/blob/d0f613fc26aafded9f40ed489a8df3dce4dfb86a/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/factories/CatalogFactoryOptions.java#L34
> 
> 
> image
> So we have Three Test case:
> 
> * TestLegacyMixedCatalog (tset case for type='arctic')
> * TestMixedCatalog containe (tset case for type='mixed_iceberg' , type='mixed_hive')
> * FlinkUnifiedCatalogITCase (tset case for type='unified')
> 
> But now, some test cases for type='mixed_iceberg' type='mixed_hive' is below TestLegacyMixedCatalog. It make confusion.
> 
> https://github.com/NetEase/amoro/blob/d0f613fc26aafded9f40ed489a8df3dce4dfb86a/mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java#L111
> 
> 
> image
> Maybe we could do some adjust in the future and make TestCase for different Catalog type more clearly @YesOrNo828 HDYT

Thank you for your clarification. We will divide it in detail in Flink's unit test compilation

**Comment by xieyi888** [comment]
> * TestLegacyMixedCatalog (tset case for type='arctic')
> * TestMixedCatalog containe (tset case for type='mixed_iceberg' , type='mixed_hive')
> * FlinkUnifiedCatalogITCase (tset case for type='unified')



> > > Thanks for your contribution.
> > > I wonder why use a legacy test to support this new feature, which doesn't quite make sense
> > 
> > 
> > It seems a little confusion As I know, LEGACY Type is for this case
> > ```
> > CREATE CATALOG <catalog_name> WITH (
> >   'type'='arctic',
> > ); 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > we will have four catalog type in the future
> > https://github.com/NetEase/amoro/blob/d0f613fc26aafded9f40ed489a8df3dce4dfb86a/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/factories/CatalogFactoryOptions.java#L34
> > 
> > image
> > So we have Three Test case:
> > 
> > * TestLegacyMixedCatalog (tset case for type='arctic')
> > * TestMixedCatalog containe (tset case for type='mixed_iceberg' , type='mixed_hive')
> > * FlinkUnifiedCatalogITCase (tset case for type='unified')
> > 
> > But now, some test cases for type='mixed_iceberg' type='mixed_hive' is below TestLegacyMixedCatalog. It make confusion.
> > https://github.com/NetEase/amoro/blob/d0f613fc26aafded9f40ed489a8df3dce4dfb86a/mixed/flink/flink-common/src/test/java/com/netease/arctic/flink/catalog/TestLegacyMixedCatalog.java#L111
> > 
> > image
> > Maybe we could do some adjust in the future and make TestCase for different Catalog type more clearly @YesOrNo828 HDYT
> 
> Thank you for your clarification. We will divide it in detail in Flink's unit test compilation

Thanks, I had mdde an issue https://github.com/NetEase/amoro/issues/2640, and I will create a PR In recent days.

**Pull request merged**